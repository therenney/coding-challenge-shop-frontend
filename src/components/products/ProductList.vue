<template>
    <div v-if="products && products.length === 0">
        <span class="uk-label uk-label-warning">No Data</span>
    </div>
    <div v-else>
        <table class="uk-table uk-table-small uk-table-responsive uk-table-striped">
            <thead>
                <tr>
                    <th class="uk-width-small">Name <span @click="sorter = 'name'"></span></th>
                    <th>Details</th>
                    <th>Price <span @click="sorter = 'price'"></span></th>
                    <th class="uk-width-small"></th>
                </tr>
            </thead>
            <tbody>
                <tr
                    v-for="{ id, name, description, manufacturer, listingPhotoUri, packaging, price } in sortedProducts"
                    :key="`product-${id}`"
                >
                    <td class="uk-text-bold">{{ name }}</td>
                    <td>
                        <div class="uk-grid-small uk-margin-bottom" uk-grid>
                            <div class="uk-width-1-1 uk-width-1-3@s">
                                <img
                                    class="product-image"
                                    :data-src="listingPhotoUri"
                                    width="100%"
                                    height="200"
                                    alt=""
                                    uk-img
                                />
                            </div>
                            <div class="uk-width-1-1 uk-width-2-3@s">
                                <dl class="uk-description-list">
                                    <dt>ID</dt>
                                    <dd>{{ id }}</dd>
                                    <dt>Manufacturer</dt>
                                    <dd>
                                        {{ manufacturer }}
                                    </dd>
                                    <dt>Packaging</dt>
                                    <dd>
                                        {{ packaging }}
                                    </dd>
                                </dl>
                            </div>
                        </div>
                    </td>
                    <td class="uk-text-bold">{{ price | currency }}</td>
                    <td>
                        <router-link class="uk-button uk-button-primary" :to="`products/${id}`">More</router-link>
                    </td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script>
export default {
    /**
     * Props definition
     * - products is required and needs to be an array
     */
    props: {
        products: {
            required: true,
            type: Array,
        },
    },

    data() {
        return {
            sortBy: 'name',
            sortDirection: 'ASC',
            search: '',
        };
    },

    computed: {
        sortedProducts() {
            return this.products;
            // const list = Object.assign({}, this.products);
            // return list.sort((a, b) => {
            //     a[this.sortBy].toLowerCase() < b[this.sortBy].toLowerCase();
            // });
        },
    },

    /**
     * Filters - Add currency filter for Table
     * Should be made globally available but for this SPA i keep it in the component
     */
    filters: {
        currency: function(value, digits) {
            if (typeof value !== 'number') {
                return value;
            }

            const formatter = new Intl.NumberFormat('de-DE', {
                style: 'currency',
                currency: 'EUR',
                minimumFractionDigits: digits,
            });

            return formatter.format(value);
        },
    },
};
</script>

<style></style>
