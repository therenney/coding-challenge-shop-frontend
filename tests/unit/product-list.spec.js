import { shallowMount } from '@vue/test-utils';
import ProductList from '@components/products/ProductList.vue';

const productData = [{
    "id": "ABC1234567800",
    "gtin": "0544900006170400",
    "manufacturer": "Beverages Ltd.",
    "name": "Beverage 23, 6x 0.75 L",
    "description": "Lorem ipsum usu amet dicat nullam ea. Nec detracto lucilius democritum in, ne usu delenit offendit deterruisset. Recusabo iracundia molestiae ea pro, suas dicta nemore an cum, erat dolorum nonummy mel ea. Iisque labores liberavisse in mei, dico laoreet elaboraret nam et, iudico verterem platonem est an. Te usu paulo vidisse epicuri, facilis mentitum liberavisse vel ut, movet iriure invidunt ut quo. Ad melius mnesarchum scribentur eum, mel at mundi impetus utroque.",
    "packaging": "CA",
    "baseProductPackaging": "BO",
    "baseProductUnit": "LT",
    "baseProductAmount": 0.75,
    "baseProductQuantity": 6,
    "price": 459,
    "logoUri": "https://dummyimage.com/200x200/eeeeee/2fd6ac.png&text=Beverages+Ltd.",
    "listingPhotoUri": "https://dummyimage.com/300x200/eeeeee/2fd6ac.png&text=Beverage+23,+6x0.75",
    "detailsPhotoUri": "https://dummyimage.com/600x400/eeeeee/2fd6ac.png&text=Beverage+23,+6x0.75"
},
{
    "id": "ABC1234567801",
    "gtin": "0544900006170401",
    "manufacturer": "Beverages Ltd.",
    "name": "Beverage 23, 12x 0.5 L",
    "description": "Viris imperdiet forensibus ius ei, ea mel modus fabellas complectitur, has decore repudiare ne. Mea graeci vivendo id, legere sententiae reprehendunt an pro. In dico quot scripta nec, pri ut ullum virtute dissentias, mel tritani officiis digniferumque at. Ad pri appareat tincidunt forensibus, cu vis omnium maluisset, nam ea dicat detraxit suavitate. Ius ei sumo aliquam takimata, mei odio graece voluptatum no, ad vel meis graecis corpora.",
    "packaging": "CA",
    "baseProductPackaging": "BO",
    "baseProductUnit": "LT",
    "baseProductAmount": 0.5,
    "baseProductQuantity": 12,
    "price": 829,
    "logoUri": "https://dummyimage.com/200x200/eeeeee/2fd6ac.png&text=Beverages+Ltd.",
    "listingPhotoUri": "https://dummyimage.com/300x200/eeeeee/2fd6ac.png&text=Beverage+23,+12x0.5",
    "detailsPhotoUri": "https://dummyimage.com/600x400/eeeeee/2fd6ac.png&text=Beverage+23,+12x0.5"
}];

describe('ProductList', () => {
    describe('Given there is no product data', () => {
        describe('When the ProductList component is rendered', () => {
            it('should render a no data message', () => {
                // Mount the ProductList
                const wrapper = shallowMount(ProductList, {
                    propsData: {
                        products: [],
                    },
                });
                expect(wrapper.find('span.uk-label').exists()).toBe(true);

            });
        });
    });

    describe('Given there is product data', () => {
        describe('When the ProductList component is rendered', () => {
            it('should render a list with all products', () => {
                // Mount the ProductList
                const wrapper = shallowMount(ProductList, {
                    propsData: {
                        products:productData,
                    },
                     stubs: {
                        'router-link': true,
                    },
                });

                // Expect that the table is rendered
                expect(wrapper.find('table').exists()).toBe(true);

                // Expect the html to match the last testrun,
                // this is a little verbose though
                expect(wrapper.html()).toMatchSnapshot();
            });
        });
    });
});
