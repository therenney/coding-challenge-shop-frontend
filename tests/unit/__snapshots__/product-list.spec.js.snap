// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Given there is product data When the ProductList component is rendered should render a list with all products 1`] = `
<div>
  <table class="uk-table uk-table-small uk-table-responsive uk-table-striped">
    <thead>
      <tr>
        <th class="uk-width-small">Name <span></span></th>
        <th>Details</th>
        <th>Price <span></span></th>
        <th class="uk-width-small"></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class="uk-text-bold">Beverage 23, 6x 0.75 L</td>
        <td>
          <div uk-grid="" class="uk-grid-small uk-margin-bottom">
            <div class="uk-width-1-1 uk-width-1-3@s"><img data-src="https://dummyimage.com/300x200/eeeeee/2fd6ac.png&amp;text=Beverage+23,+6x0.75" width="100%" height="200" alt="" uk-img="" class="product-image"></div>
            <div class="uk-width-1-1 uk-width-2-3@s">
              <dl class="uk-description-list">
                <dt>ID</dt>
                <dd>ABC1234567800</dd>
                <dt>Manufacturer</dt>
                <dd>
                  Beverages Ltd.
                </dd>
                <dt>Packaging</dt>
                <dd>
                  CA
                </dd>
              </dl>
            </div>
          </div>
        </td>
        <td class="uk-text-bold">€459.00</td>
        <td>
          <router-link-stub to="products/ABC1234567800" class="uk-button uk-button-primary">More</router-link-stub>
        </td>
      </tr>
      <tr>
        <td class="uk-text-bold">Beverage 23, 12x 0.5 L</td>
        <td>
          <div uk-grid="" class="uk-grid-small uk-margin-bottom">
            <div class="uk-width-1-1 uk-width-1-3@s"><img data-src="https://dummyimage.com/300x200/eeeeee/2fd6ac.png&amp;text=Beverage+23,+12x0.5" width="100%" height="200" alt="" uk-img="" class="product-image"></div>
            <div class="uk-width-1-1 uk-width-2-3@s">
              <dl class="uk-description-list">
                <dt>ID</dt>
                <dd>ABC1234567801</dd>
                <dt>Manufacturer</dt>
                <dd>
                  Beverages Ltd.
                </dd>
                <dt>Packaging</dt>
                <dd>
                  CA
                </dd>
              </dl>
            </div>
          </div>
        </td>
        <td class="uk-text-bold">€829.00</td>
        <td>
          <router-link-stub to="products/ABC1234567801" class="uk-button uk-button-primary">More</router-link-stub>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
